src/doc/trpl/the-stack-and-the-heap.md

```diff
@@ -1,3 +1,526 @@
+
+| Address | Name | Value |
++---------+------+-------+
+| 3       | c    | 1     |
+| 2       | b    | 100   |
+| 1       | a    | 5     |
+| 0       | x    | 42    |
+
+And then `foo()` calls `bar()`:
+
+| Address | Name | Value |
++---------+------+-------+
+| 4       | i    | 6     |
+| 3       | c    | 1     |
+| 2       | b    | 100   |
+| 1       | a    | 5     |
+| 0       | x    | 42    |
+
+Whew! Our stack is growing tall.
+
+After `bar()` is over, its frame is deallocated, leaving just `foo()` and
+`main()`:
+
+| Address | Name | Value |
++---------+------+-------+
+| 2       | b    | 100   |
+| 1       | a    | 5     |
+| 0       | x    | 42    |
+
+And then `foo()` ends, leaving just `main()`
+
+| Address | Name | Value |
++---------+------+-------+
+| 0       | x    | 42    |
+
+And then we’re done. Getting the hang of it? It’s like piling up dishes: you
+add to the top, you take away from the top.
+
+# The Heap
+
+Now, this works pretty well, but not everything can work like this. Sometimes,
+you need to pass some memory between different functions, or keep it alive for
+longer than a single function’s execution. For this, we can use the heap.
+
+In Rust, you can allocate memory on the heap with the [`Box<T>` type][box].
+Here’s an example:
+
+```rust
+fn main() {
+    let x = Box::new(5);
+    let y = 42;
+}
+```
+
+[box]: ../std/boxed/index.html
+
+Here’s what happens in memory when `main()` is called:
+
+| Address | Name | Value  |
++---------+------+--------+
+| 1       | y    | 42     |
+| 0       | x    | ?????? |
+
+We allocate space for two variables on the stack. `y` is `42`, as it always has
+been, but what about `x`? Well, `x` is a `Box<i32>`, and boxes allocate memory
+on the heap. The actual value of the box is a structure which has a pointer to
+‘the heap’. When we start executing the function, and `Box::new()` is called,
+it allocates some memory for the heap, and puts `5` there. The memory now looks
+like this:
+
+| Address | Name | Value  |
++---------+------+--------+
+| 2 ^ 30  |      | 5      |
+| ...     | ...  | ...    |
+| 1       | y    | 42     |
+| 0       | x    | 2 ^ 30 |
+
+We have `2 ^ 30` in our hypothetical computer with 1GB of RAM. And since our
+stack grows from zero, the easiest place to allocate memory is from the other
+end. So our first value is at the highest place in memory. And the value of the
+struct at `x` has a [raw pointer][rawpointer] to the place we’ve allocated on
+the heap, so the value of `x` is `2 ^ 30`, the memory location we’ve asked for.
+
+[rawpointer]: raw-pointers.html
+
+Since this memory is on the heap, it can stay alive longer than the function
+which allocates the box. In this case, however, it doesn’t. When the function
+is over, we need to free the stack frame for `main()`. `Box<T>`, though, has a
+trick up its sleve: [Drop][drop]. The implementation of `Drop` for `Box`
+deallocates the memory that was allocated when it was created. Great! So when
```

Comment:
you haven't really said what it means to "deallocate" memory with respect to the heap.  Based on the initial presentation where you said it comes from the other end of memory, one might think that the heap is merely another stack that just grows in the opposite direction, and miss the detail that the runtime system is going to allow memory to be freed in non-LIFO order.